libs/nim/verify-utils/verify_given_proof_constantine.nim

```diff
@@ -0,0 +1,218 @@
+import
+  nimcrypto/[sha2, hash, utils],
+  stew/byteutils,
+  std/[strutils,json]
+
+import # constantine imports
+  constantine/math/pairings/pairings_bn,
+  tests/math/t_pairing_template,
+  constantine/math/io/[io_ec, io_fields, io_bigints],
+  constantine/math/elliptic/ec_scalar_mul,
+  constantine/math/config/type_bigint
+
+type
+  IC* = array[5, ECP_ShortW_Aff[Fp[BN254_Snarks], G1]]
+
+  VerificationKey* = object
+    alpha*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+    beta*, gamma*, delta*: ECP_ShortW_Aff[Fp2[BN254_Snarks], G2]
+    ic*: IC
+
+  Proof* = object
+    a*, c*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+    b*: ECP_ShortW_Aff[Fp2[BN254_Snarks], G2]
+
+  Header* = object
+    head*: Fr[BN254_Snarks]
+    tail*: Fr[BN254_Snarks]
+
+  Input* = object
+    data*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+
+proc concatArrays*(
+    currentHeaderHash: array[32, byte], newOptimisticHeader: array[32, byte],
+    newFinalizedHeader: array[32, byte], newExecutionStateRoot: array[32, byte],
+    zerosSlotBuffer: array[24, byte], currentSlot: array[8, byte],
+    domain: array[32, byte]): array[192, byte] =
+  var res: array[192, byte]
+  res[0..31] = currentHeaderHash
+  res[32..63] = newOptimisticHeader
+  res[64..95] = newFinalizedHeader
+  res[96..127] = newExecutionStateRoot
+  res[128..151] = zerosSlotBuffer
+  res[152..159] = currentSlot
+  res[160..191] = domain
+
+  res
+
+proc hashHeaders*(
+    currentHeaderHash: array[32, byte], newOptimisticHeader: array[32, byte],
+    newFinalizedHeader: array[32, byte], newExecutionStateRoot: array[32, byte],
+    zerosSlotBuffer: array[24, byte], currentSlot: array[8, byte],
+    domain: array[32, byte]): array[32, byte]  =
+  let concat = (concatArrays(currentHeaderHash, newOptimisticHeader,
+                             newFinalizedHeader, newExecutionStateRoot,
+                             zerosSlotBuffer, currentSlot, domain))
+
+
+  let hash = sha2.sha256.digest(concat)
+  hash.data
+
+proc TwoOnPower*(power: int): int =
+  var output = 1
+  for i in 1..power:
+    output *= 2
+  output
```

Comment:
Do we really need this? Can't we use the [^ operator](https://nim-lang.org/docs/math.html#%5E%2CT%2CNatural)
```nim
import std/math
let a =  2^5
assert(a == 32)
```